subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/FileParameterUtils.java

```diff
@@ -135,11 +136,11 @@ private static void resolveCompositeOutputFilePropertySpecs(final String ownerDi
                     throw new IllegalArgumentException(String.format("Mapped output property '%s' has null key", propertyName));
                 }
                 String id = key.toString();
-                FileCollectionInternal outputFiles = fileCollectionFactory.resolving(entry.getValue());
+                FileCollectionInternal outputFiles = fileCollectionFactory.resolving(AbsentProviderHandling.REQUIRE_PRESENT, entry.getValue());
                 consumer.accept(new DefaultCacheableOutputFilePropertySpec(propertyName, "." + id, outputFiles, outputType));
             }
         } else {
-            FileCollectionInternal outputFileCollection = fileCollectionFactory.resolving(unpackedValue);
+            FileCollectionInternal outputFileCollection = fileCollectionFactory.resolving(AbsentProviderHandling.REQUIRE_PRESENT, unpackedValue);
             AtomicInteger index = new AtomicInteger(0);
             outputFileCollection.visitStructure(new FileCollectionStructureVisitor() {
                 @Override
```

Comment:
What strikes me here is that all the calls to `fileCollectionFactory.resolving()` in `resolveInputFileValue()` take `ALLOW_ABSENT`, while all the calls in `resolveOutputFilePropertySpecs()` take `REQUIRE_PRESENT`. I wonder if we could somehow remove this redundancy. Something like `fileCollectionFactory.lenient()` giving you a variant of the factory that uses `ALLOW_ABSENT` perhaps?

Reply:
I added `resolvingIfPresent` to `FileCollectionFactory` and removed the duplication from the to `resolve` methods in `FileParametersUtils`. I think it is clearer now. I don't want to create a full `FileCollectionFactory` which has a lot of lenient methods. So I would need to introduce a new interface with only two methods.